source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc

```diff
@@ -123,15 +124,137 @@ AWSLambdaFilter::decodeHeaders(Http::RequestHeaderMap &headers,
 
 
 Http::FilterHeadersStatus 
-AWSLambdaFilter::encodeHeaders(Http::ResponseHeaderMap &headers, bool ) {
+AWSLambdaFilter::encodeHeaders(Http::ResponseHeaderMap &headers, bool) {
 
   if (!headers.get(AWSLambdaHeaderNames::get().FunctionError).empty()){
     // We treat upstream function errors as if it was any other upstream error
     headers.setStatus(504);
   }
+  response_headers_ = &headers;
+  if (functionOnRoute() != nullptr && functionOnRoute()->unwrapAsAlb()){
+    // Stop iteration so that encodedata can mutate headers from alb json
+    return Http::FilterHeadersStatus::StopIteration;
+  }
   return Http::FilterHeadersStatus::Continue;
 }
 
+Http::FilterDataStatus AWSLambdaFilter::encodeData(
+                                      Buffer::Instance &data, bool end_stream ){  
+
+  if (state_ == State::Destroyed){
+    // Safety against use after free if we exceed buffer limit 
+    // during modifications. This should never happen as we only shrink.
+    ENVOY_LOG(debug, "{}: attempted operations while destroyed", __func__);
+    return Http::FilterDataStatus::StopIterationNoBuffer;
+  }
+
+  if (functionOnRoute() == nullptr || !functionOnRoute()->unwrapAsAlb()){
+    // return response as is if not configured for alb mode
+    return Http::FilterDataStatus::Continue;
+  }
+
+  // support for streaming and http2 for future proofing
+  encoder_callbacks_->addEncodedData(data, false);
+
+  if (!end_stream) {
+    // we need the entire response prior to parsing and unwrapping the json
+    return Http::FilterDataStatus::StopIterationAndBuffer;
+  }
+
+  // Now that the response is finished we know that the following is safe
+  // as the following options will only make the resulting buffer smaller.
+  const Buffer::Instance&  buff = *encoder_callbacks_->encodingBuffer();
+  encoder_callbacks_->modifyEncodingBuffer([this](Buffer::Instance& enc_buf) {
+    Buffer::OwnedImpl body;
+    parseResponseAsALB(*response_headers_, enc_buf, body);
+    enc_buf.drain(enc_buf.length());
+    enc_buf.move(body);
+  });
+  response_headers_->setContentLength(buff.length());
```

Comment:
i think this can be extracted to a helper function and re-used in encode trailers?

Reply:
Set in finalize()